client/dashboard/domains/domain-contact-details/region-address-fieldsets.tsx

```diff
@@ -0,0 +1,197 @@
+import { type StatesListItem, type DomainContactDetails } from '@automattic/api-core';
+import { __experimentalInputControl as InputControl, SelectControl } from '@wordpress/components';
+import { useViewportMatch } from '@wordpress/compose';
+import { type Field, type DataFormControlProps } from '@wordpress/dataviews';
+import { __ } from '@wordpress/i18n';
+import { useEffect } from 'react';
+import { getCountryPostalCodeSupport } from './custom-form-fieldsets/get-country-postal-code-support';
+import {
+	CHECKOUT_EU_ADDRESS_FORMAT_COUNTRY_CODES,
+	CHECKOUT_UK_ADDRESS_FORMAT_COUNTRY_CODES,
+	type CountryListItem,
+} from './custom-form-fieldsets/types';
+
+const getPostalCodeLabel = ( countryCode: string ): string => {
+	switch ( countryCode ) {
+		case 'US':
+			return __( 'ZIP code' );
+		default:
+			return __( 'Postal Code' );
+	}
+};
+
+const getStateSelectLabel = (
+	countryCode: string,
+	statesList: StatesListItem[] | undefined
+): string => {
+	switch ( countryCode ) {
+		case 'CA':
+			return __( 'Select Province' );
+		default:
+			if ( statesList && statesList.length > 0 ) {
+				return __( 'Select State' );
+			}
+			return __( 'State' );
+	}
+};
+
+const createStateFieldEdit = ( statesList: StatesListItem[] | undefined, countryCode: string ) => {
+	const StateFieldEdit = ( {
+		field,
+		onChange,
+		data,
+		hideLabelFromVision,
+	}: DataFormControlProps< DomainContactDetails > ) => {
+		const { id, getValue } = field;
+		const currentValue = getValue?.( { item: data } );
+
+		// If the item data is not in the statesList, set the state to the first option
+		useEffect( () => {
```

Comment:
As always, I find this sort of `useEffect` suspect 😅 But I admit changing it would require a bit of refactor. The check would need to move up into the `<ContactForm>` component which owns both the `formData` and the `statesList`.

Reply:
Sounds good, [I fixed it here](https://github.com/Automattic/wp-calypso/pull/106157/commits/df31cdb1848b38ddfb9b66bf2181ba8bb4ccaa30).